File: packages/zynth-core/ios/ZynthKit/src/runtime/WebSocketModule.swift
import Foundation

@objc(WebSocketModule)
final class WebSocketModule: NSObject, ZynthModule, URLSessionWebSocketDelegate {
  let name: String = "WebSocket"
  private weak var runtime: ZynthRuntime?
  private var session: URLSession!
  private var sockets: [Int: URLSessionWebSocketTask] = [:]
  private let queue = DispatchQueue(label: "dev.zynth.websocket")

  init(runtime: ZynthRuntime) {
    self.runtime = runtime
    super.init()
    session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
  }

  var exportedMethods: [String] {
    ["connect", "send", "close", "reloadDevBundle"]
  }

  func invalidate() {
    queue.sync {
      for socket in sockets.values {
        socket.cancel(with: .goingAway, reason: nil)
      }
      sockets.removeAll()
    }
    session.invalidateAndCancel()
  }

  func call(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "connect":
      return try connect(args)
    case "send":
      return try send(args)
    case "close":
      return try close(args)
    case "reloadDevBundle":
      return reloadDevBundle(args)
    default:
      return ["error": "unknown_method", "method": method]
    }
  }

  private func connect(_ args: ZynthArgs) throws -> Any {
    let id = Int(try args.number("id"))
    let urlString = try args.string("url")
    guard let url = URL(string: urlString) else {
      return ["error": "invalid_url", "message": urlString]
    }
    let task = session.webSocketTask(with: url)
    queue.sync {
      sockets[id] = task
    }
    task.resume()
    receiveLoop(id: id, task: task)
    return ["result": true]
  }

  private func send(_ args: ZynthArgs) throws -> Any {
    let id = Int(try args.number("id"))
    let data = try args.string("data")
    guard let task = queue.sync(execute: { sockets[id] }) else {
      return ["error": "socket_not_found"]
    }
    task.send(.string(data)) { [weak self] error in
      if let error {
        self?.emit(id: id, type: "error", ["message": error.localizedDescription])
      }
    }
    return ["result": true]
  }

  private func close(_ args: ZynthArgs) throws -> Any {
    let id = Int(try args.number("id"))
    let codeValue = Int(args.number("code", default: 1000))
    let reason = args.string("reason", default: "")
    let closeCode = URLSessionWebSocketTask.CloseCode(rawValue: codeValue) ?? .normalClosure
    let reasonData = reason.isEmpty ? nil : reason.data(using: .utf8)
    let task = queue.sync {
      sockets.removeValue(forKey: id)
    }
    task?.cancel(with: closeCode, reason: reasonData)
    guard task != nil else { return ["result": true] }
    emit(id: id, type: "close", ["code": codeValue, "reason": reason, "wasClean": true])
    return ["result": true]
  }

  private func reloadDevBundle(_ args: ZynthArgs) -> Any {
    guard let runtime else {
      return ["error": "runtime_deallocated"]
    }
    DispatchQueue.main.async {
      runtime.refreshDevBundle()
    }
    return ["result": true]
  }

  private func receiveLoop(id: Int, task: URLSessionWebSocketTask) {
    task.receive { [weak self] result in
      guard let self else { return }
      switch result {
      case .success(let message):
        switch message {
        case .string(let text):
          self.emit(id: id, type: "message", ["data": text])
        case .data:
          self.emit(id: id, type: "error", ["message": "Binary WebSocket frames are not supported"])
        @unknown default:
          self.emit(id: id, type: "error", ["message": "Unknown WebSocket frame"])
        }
        if self.queue.sync(execute: { self.sockets[id] != nil }) {
          self.receiveLoop(id: id, task: task)
        }
      case .failure(let error):
        let task = self.queue.sync {
          self.sockets.removeValue(forKey: id)
        }
        guard task != nil else { return }
        self.emit(id: id, type: "error", ["message": error.localizedDescription])
        self.emit(id: id, type: "close", ["code": 1006, "reason": error.localizedDescription, "wasClean": false])
      }
    }
  }

  private func emit(id: Int, type: String, _ data: [String: Any] = [:]) {
    var payload = data
    payload["id"] = id
    payload["type"] = type
    runtime?.emitEvent(name: "zynth.websocket.event", payload: payload)
  }

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didOpenWithProtocol protocol: String?
  ) {
    let id = queue.sync { sockets.first(where: { $0.value == webSocketTask })?.key }
    guard let id else { return }
    emit(id: id, type: "open")
  }

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
    reason: Data?
  ) {
    let id = queue.sync { sockets.first(where: { $0.value == webSocketTask })?.key }
    guard let id else { return }
    let task = queue.sync {
      sockets.removeValue(forKey: id)
    }
    guard task != nil else { return }
    let reasonText = reason.flatMap { String(data: $0, encoding: .utf8) } ?? ""
    emit(id: id, type: "close", ["code": closeCode.rawValue, "reason": reasonText, "wasClean": closeCode == .normalClosure])
  }
}

File: packages/zynth-core/android/ZynthKit/src/main/java/com/zynth/kit/runtime/modules/WebSocketModule.kt
package com.zynth.kit.runtime.modules

import com.zynth.kit.runtime.ZynthArgs
import com.zynth.kit.runtime.ZynthModule
import com.zynth.kit.runtime.ZynthRuntime
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

internal class WebSocketModule(private val runtime: ZynthRuntime) : ZynthModule {
  override val name: String = "WebSocket"
  override val exportedMethods: List<String> = listOf(
    "connect",
    "send",
    "close",
    "reloadDevBundle",
  )

  private val client = OkHttpClient.Builder()
    .retryOnConnectionFailure(true)
    .pingInterval(30, TimeUnit.SECONDS)
    .build()
  private val sockets = ConcurrentHashMap<Int, WebSocket>()

  override fun invalidate() {
    sockets.forEach { (_, socket) ->
      socket.close(1001, "runtime invalidated")
    }
    sockets.clear()
  }

  override fun call(method: String, args: ZynthArgs): JSONObject {
    return when (method) {
      "connect" -> connect(args)
      "send" -> send(args)
      "close" -> close(args)
      "reloadDevBundle" -> reloadDevBundle(args)
      else -> JSONObject().put("error", "unknown_method").put("method", method)
    }
  }

  private fun connect(args: ZynthArgs): JSONObject {
    val id = args.getInt("id")
    val url = args.getString("url")
    val request = Request.Builder().url(url).build()
    client.newWebSocket(request, object : WebSocketListener() {
      override fun onOpen(webSocket: WebSocket, response: Response) {
        sockets[id] = webSocket
        emit(id, "open")
      }

      override fun onMessage(webSocket: WebSocket, text: String) {
        emit(id, "message", JSONObject().put("data", text))
      }

      override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
        webSocket.close(code, reason)
      }

      override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
        sockets.remove(id)
        emit(id, "close", JSONObject()
          .put("code", code)
          .put("reason", reason)
          .put("wasClean", code == 1000)
        )
      }

      override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        sockets.remove(id)
        emit(id, "error", JSONObject().put("message", t.message ?: "WebSocket error"))
        emit(id, "close", JSONObject()
          .put("code", 1006)
          .put("reason", t.message ?: "WebSocket failure")
          .put("wasClean", false)
        )
      }
    })
    return JSONObject().put("result", true)
  }

  private fun send(args: ZynthArgs): JSONObject {
    val id = args.getInt("id")
    val data = args.getString("data")
    val socket = sockets[id] ?: return JSONObject().put("error", "socket_not_found")
    val ok = socket.send(data)
    return JSONObject().put("result", ok)
  }

  private fun close(args: ZynthArgs): JSONObject {
    val id = args.getInt("id")
    val code = args.getInt("code", 1000)
    val reason = args.getString("reason", "")
    val socket = sockets.remove(id) ?: return JSONObject().put("result", true)
    socket.close(code, reason)
    return JSONObject().put("result", true)
  }

  private fun reloadDevBundle(args: ZynthArgs): JSONObject {
    val baseUrl = args.getOptionalString("url")
      ?: return JSONObject().put("error", "invalid_url")
    val token = args.getOptionalString("token")
    val rootId = args.getInt("rootId", runtime.rootSurfaceId)
    Thread({
      try {
        val base = baseUrl.toHttpUrlOrNull() ?: throw IllegalArgumentException("Invalid dev server URL")
        val builder = base.newBuilder().encodedPath("/main.js")
        if (!token.isNullOrBlank()) {
          builder.removeAllQueryParameters("token")
          builder.addQueryParameter("token", token)
        }
        val request = Request.Builder()
          .url(builder.build())
          .header("Cache-Control", "no-cache")
          .build()
        client.newCall(request).execute().use { response ->
          if (!response.isSuccessful) {
            throw IllegalStateException("Bundle reload failed with ${response.code}")
          }
          val code = response.body?.string() ?: throw IllegalStateException("Bundle reload returned empty body")
          runtime.evaluateScript(code, "main.js")
          runtime.evaluateScript("globalThis.__startApp && globalThis.__startApp($rootId);", "dev-bundle-restart.js")
        }
      } catch (t: Throwable) {
        emit(-1, "error", JSONObject().put("message", "Dev bundle reload failed: ${t.message}"))
      }
    }, "ZynthDevBundleReload").start()
    return JSONObject().put("result", true)
  }

  private fun emit(id: Int, type: String, payload: JSONObject = JSONObject()) {
    payload.put("id", id)
    payload.put("type", type)
    runtime.emitEvent("zynth.websocket.event", payload)
  }
}
